client/blocks/upload-image/index.jsx

```diff
@@ -48,51 +69,145 @@ class UploadImage extends Component {
+			selectedImage: imageObjectUrl,
+			selectedImageName: fileName,
 		} );
 	};
 
-	onImageEditorDone = ( error, imageBlob ) => {
-		this.hideImageEditor();
+	handleError = ( errorConst, error = '' ) => {
+		const { onError } = this.props;
+
+		let message = error;
+
+		if ( errorConst === ERROR_UPLOADING_IMAGE ) {
+			if ( error.length && error[ 0 ] === ValidationErrors.SERVER_ERROR ) {
+				message = ERROR_STRINGS[ ValidationErrors.SERVER_ERROR ]();
+			}
+		}
+
+		onError( errorConst, message );
+	};
 
+	onImageEditorDone = ( error, imageBlob, imageEditorProps ) => {
 		if ( error ) {
-			return;
+			this.handleError( ERROR_IMAGE_EDITOR_DONE, error );
 		}
 
-		this.setState( { editedImage: URL.createObjectURL( imageBlob ) } );
+		this.setState( {
+			editedImage: URL.createObjectURL( imageBlob ),
+			isUploading: true,
+		} );
+
+		MediaStore.off( 'change', this.handleMediaStoreChange );
+		MediaValidationStore.off( 'change', this.storeValidationErrors );
+
+		this.props.onImageEditorDone( imageBlob, imageEditorProps );
+
+		this.uploadImage( imageBlob, imageEditorProps );
+
+		this.hideImageEditor();
+	};
+
+	uploadImage( imageBlob, imageEditorProps ) {
+		const { siteId } = this.props;
+
+		const { fileName, mimeType } = imageEditorProps;
+
+		const transientImageId = uniqueId( 'upload-image-' );
+
+		this.uploadingImageTransientId = transientImageId;
+
+		const item = {
+			ID: transientImageId,
+			fileName: fileName,
+			fileContents: imageBlob,
+			mimeType: mimeType,
+		};
+
+		MediaValidationStore.on( 'change', this.storeValidationErrors );
+
+		MediaStore.on( 'change', this.handleMediaStoreChange );
 
-		this.props.onImageEditorDone( imageBlob );
+		// Upload the image.
+		MediaActions.add( siteId, item );
+
+		this.setState( { isUploading: true } );
+	}
+
+	// Handle the uploading process.
+	handleMediaStoreChange = () => {
+		const { siteId, onUploadImageDone } = this.props;
+
+		const { errors } = this.state;
+
+		const uploadedImage = MediaStore.get( siteId, this.uploadingImageTransientId );
+		const isUploadInProgress = uploadedImage && MediaUtils.isItemBeingUploaded( uploadedImage );
+
+		// File has finished uploading or failed.
+		if ( ! isUploadInProgress ) {
+			this.setState( {
+				isUploading: false,
+			} );
+
+			if ( uploadedImage && uploadedImage.URL ) {
+				this.uploadingImageTransientId = null;
+
+				this.setState( { uploadedImage: uploadedImage } );
+
+				MediaStore.off( 'change', this.handleMediaStoreChange );
+
+				onUploadImageDone( uploadedImage );
+
+				return;
+			}
+
+			const validationErrors = errors[ this.uploadingImageTransientId ] || [];
+			this.handleError( ERROR_UPLOADING_IMAGE, validationErrors );
+		}
+	};
+
+	storeValidationErrors = () => {
+		this.setState( {
+			errors: MediaValidationStore.getAllErrors( this.props.siteId ),
+		} );
 	};
 
 	hideImageEditor = () => {
 		const { resetAllImageEditorState: resetAllImageEditorStateAction } = this.props;
 
 		resetAllImageEditorStateAction();
```

Comment:
I don't think "mentioning" the action in `connect` is causing trouble. That's what `connect` is for. :) But sure, I can call it like `this.props.resetAllImageEditorStateAction`. :+1: